Fix 502 errors when Nginx is running before SPN

citronalco opened the pull request:
Fixes 502 errors when Nginx is running before SPN

If Nginx can't resolve an upstream server, it stops trying to resolve it via IPv4 very quickly and continues using only IPv6.
As the website is only available with IPv4 on port 8000, this fails.

citronalco committed:
Nginx: only resolve SPN components with IPv4
Fixes 502 errors when Nginx is running before SPN

cfr34k commented:
If I understand your change correctly, it also configures nginx to use a DNS resolver on `localhost`. I think this might be problematic on some systems (especially if they don't use Network Manager or something similar).

I'd suggest to simply use `proxy_pass http://127.0.0.1:8000` with the IP address instead of `localhost`. What do you think?

citronalco committed:
Update nginx.conf

- use resolver only for "location /" which does neither work with ipv6 nor with 127.0.0.1
- switch websocket from localhost to 127.0.0.1

citronalco commented:
You're right, with this change nginx tries to resolve names via localhost. Which should not be a problem, as the only name to resolve is "localhost".

For some reason using 127.0.0.1 instead of localhost works only for the websocket, but not for the website on port 8000. I get "400 - Bad request" then.

I've updated the PR to use 127.0.0.1 at least for the websocket, and moved the "resolver" line into the website's "location" definition.

cfr34k commented:
Thanks, I can accept it this way.

However, something feels wrong about the Bad Request error, because the website/Django actually listens on 127.0.0.1:8000 (you can check with `ss -tpn`). Did you find any hints in the logs of nginx or did Django print something useful on the terminal about why this happens?

citronalco commented:
There's no error in Nginx's log file, so I think the "Bad Request (400)" comes directly from website/Django, though it does not print anything at all.

I have no clue about Django, so if you want some debug logs you'll have to tell me how to do that.

cfr34k commented:
Normally Django prints errors to stderr. If your system is protected, you can also enable Django’s debugging in `website/Programmierspiel/local_settings.py` and you should geht a more informative error page.

citronalco commented:
Indeed, it's more informative:

> DisallowedHost at /watch/
>
> Invalid HTTP_HOST header: '127.0.0.1:8000'. You may need to add '127.0.0.1' to ALLOWED_HOSTS.

After changing `ALLOWED_HOSTS` in website/Programmierspiel/local_settings.py from ['localhost'] to ['127.0.0.1'] it works.

citronalco committed:
README: Do not set Django's ALLOWED_HOSTS to localhost but to
127.0.0.1
nginx.conf: Use 127.0.0.1 instead of localhost for the website

citronalco commented:
In case you did not notice: I modified the pull request, should be ok now.

cfr34k commented:
Thanks for the update!
